lightning-block-sync/src/lib.rs

```diff
@@ -124,3 +138,563 @@
+	DisconnectBlock(ValidatedBlockHeader),
+	ConnectBlock(ValidatedBlockHeader),
+}
+
+impl<C: Cache> ChainNotifier<C> {
+	/// Finds the fork point between `new_header` and `old_header`, disconnecting blocks from
+	/// `old_header` to get to that point and then connecting blocks until `new_header`.
+	///
+	/// Validates headers along the transition path, but doesn't fetch blocks until the chain is
+	/// disconnected to the fork point. Thus, this may return an `Err` that includes where the tip
+	/// ended up which may not be `new_header`. Note that iff the returned `Err` contains `Some`
+	/// header then the transition from `old_header` to `new_header` is valid.
+	async fn sync_listener<L: ChainListener, P: Poll>(
+		&mut self,
+		new_header: ValidatedBlockHeader,
+		old_header: &ValidatedBlockHeader,
+		chain_poller: &mut P,
+		chain_listener: &mut L,
+	) -> Result<(), (BlockSourceError, Option<ValidatedBlockHeader>)> {
+		let mut events = self.find_fork(new_header, old_header, chain_poller).await.map_err(|e| (e, None))?;
+
+		let mut last_disconnect_tip = None;
+		let mut new_tip = None;
+		for event in events.iter() {
+			match &event {
+				&ForkStep::DisconnectBlock(ref header) => {
+					println!("Disconnecting block {}", header.block_hash);
+					if let Some(cached_header) = self.header_cache.remove(&header.block_hash) {
+						assert_eq!(cached_header, *header);
+					}
+					chain_listener.block_disconnected(&header.header, header.height);
+					last_disconnect_tip = Some(header.header.prev_blockhash);
+				},
+				&ForkStep::ForkPoint(ref header) => {
+					new_tip = Some(*header);
+				},
+				_ => {},
+			}
+		}
+
+		// If blocks were disconnected, new blocks will connect starting from the fork point.
+		// Otherwise, there was no fork, so new blocks connect starting from the old tip.
+		assert_eq!(last_disconnect_tip.is_some(), new_tip.is_some());
+		if let &Some(ref tip_header) = &new_tip {
+			debug_assert_eq!(tip_header.header.block_hash(), *last_disconnect_tip.as_ref().unwrap());
+		} else {
+			new_tip = Some(*old_header);
+		}
+
+		for event in events.drain(..).rev() {
+			if let ForkStep::ConnectBlock(header) = event {
+				let block = chain_poller
+					.fetch_block(&header).await
+					.or_else(|e| Err((e, new_tip)))?;
+				debug_assert_eq!(block.block_hash, header.block_hash);
+
+				println!("Connecting block {}", header.block_hash);
+				self.header_cache.insert(header.block_hash, header);
+				chain_listener.block_connected(&block, header.height);
+				new_tip = Some(header);
+			}
+		}
+		Ok(())
+	}
+
+	/// Walks backwards from `current_header` and `prev_header`, finding the common ancestor.
+	/// Returns the steps needed to produce the chain with `current_header` as its tip from the
+	/// chain with `prev_header` as its tip. There is no ordering guarantee between different
+	/// `ForkStep` types, but `DisconnectBlock` and `ConnectBlock` are each returned in
+	/// height-descending order.
+	async fn find_fork<P: Poll>(
```

Comment:
please see if you like this simplification: https://github.com/rust-bitcoin/rust-lightning/commit/9e07f878f9601340ac050f47b62fdd6f908fdc89